src/network/networkd-link.c

```diff
@@ -3114,6 +3117,9 @@ static int link_carrier_lost(Link *link) {
         if (link->setting_mtu)
                 return 0;
 
+        if (link->state == LINK_STATE_CONFIGURED && link->network->ignore_carrier_gainloss)
+                return 0;
+
         r = link_stop_clients(link);
```

Comment:
The only change required for this feature is 

```
git diff
diff --git a/src/network/networkd-link.c b/src/network/networkd-link.c
index 67cdea0b9..3c24167cb 100644
--- a/src/network/networkd-link.c
+++ b/src/network/networkd-link.c
@@ -3123,9 +3123,11 @@ static int link_carrier_lost(Link *link) {
         if (link_dhcp4_server_enabled(link))
                 (void) sd_dhcp_server_stop(link->dhcp_server);
 
-        r = link_drop_config(link);
-        if (r < 0)
-                return r;
+        if (link->state == LINK_STATE_CONFIGURED && !link->network->ignore_carrier_gainloss) {
+                r = link_drop_config(link);
+                if (r < 0)
+                        return r;
+        }
```

Reply:
or could be just 
```
       if (!link->network->ignore_carrier_gainloss) {
+                r = link_drop_config(link);
+                if (r < 0)
+                        return r;
+        }
```

Reply:
But then it will reach line 3138 which will call  link_drop_foreign_config() which in turn calls address_remove() and route_remove(),

Now **I do not know what foreign config is** (Please shed some light if you know) but do we need to remove address and route for foreign config or not?

My main purpose is that addresses and routes should not be removed just because of lost carrier.

Reply:
`networkd` either manages a interface fully or not. So while running `networkd` we can also add address/routes via different tools such is iproute. When that happens kernel pushes those via netlink to `networkd` and we store them on foreign. In short configs which are not done by networkd.

Reply:
Exactly - so we should not delete those foreign routes and addresses too... if networkd has not added them then why delete them?

This was the initial reason I wanted this IgnoreCarrierGainLoss= feature.

In short when carrier is lost - do not delete any route / addresses. (and do not add them again either when carrier is back)

So I believe my modifications are more appropriate. (it ignores DHCP / non-static config)

Hope you agree and we come to some conclusion.

Reply:
No Again networkd either manages the interface fully or not. If you want to configure the interface with another tool why use networkd . This will make lots of mess.

Reply:
Just forgot If some feature is missing with networkd that makes you use another tool then Please open a RFE . We will try to address it .

Reply:
No my point here is... if I set IgnoreCarrierGainLoss=true then it should not delete self-managed addresses routes as well as foreign addresses / routes.

Changes that you suggested - does not delete self-managed routes but it goes ahead and deletes foreign addresses / routes. It should retain that too.

Reply:
>No my point here is... if I set IgnoreCarrierGainLoss=true then it should not delete self-managed addresses routes as well as foreign addresses / routes.
> Changes that you suggested - does not delete self-managed routes but it goes ahead and deletes foreign addresses / routes. It should retain that too.

Again networkd either manages the interface fully or not.

Reply:
Yes I understand that networkd wants to manage fully but when IgnoreCarrierGainLoss=true is set - why is it necessary to delete foreign config?

If we are keeping self-managed config whats the harm in keeping foreign config as well? Why drop it?

Please think from static address point of view not DHCP (as my code already excludes DHCP)

Idea is not to do ANYTHING when carrier is lost (and regained) but for some reason you are insisting for static addresses, managed config will be kept but foreign will be dropped.

We need someone else's view too. May be @poettering can you please look? Thanks.